packages/server/graphql/public/mutations/acceptRequestToJoinDomain.ts

```diff
@@ -0,0 +1,131 @@
+import {getUserId} from '../../../utils/authorization'
+import {MutationResolvers} from '../resolverTypes'
+import getKysely from '../../../postgres/getKysely'
+import standardError from '../../../utils/standardError'
+import getRethink from '../../../database/rethinkDriver'
+import getTeamsByIds from '../../../postgres/queries/getTeamsByIds'
+import RedisLock from '../../../utils/RedisLock'
+import insertNewTeamMember from '../../../safeMutations/insertNewTeamMember'
+import addTeamIdToTMS from '../../../safeMutations/addTeamIdToTMS'
+import adjustUserCount from '../../../billing/helpers/adjustUserCount'
+import {InvoiceItemType, SubscriptionChannel} from 'parabol-client/types/constEnums'
+import setUserTierForUserIds from '../../../utils/setUserTierForUserIds'
+import publish from '../../../utils/publish'
+import toTeamMemberId from 'parabol-client/utils/relay/toTeamMemberId'
+
+// TODO (EXPERIMENT: prompt-to-join-org): some parts are borrowed from acceptTeamInvitation, create generic functions
+const acceptRequestToJoinDomain: MutationResolvers['acceptRequestToJoinDomain'] = async (
+  _source,
+  {requestId, teamIds},
+  {authToken, dataLoader, socketId: mutatorId}
+) => {
+  const operationId = dataLoader.share()
+  const subOptions = {mutatorId, operationId}
+  const viewerId = getUserId(authToken)
+  const now = new Date()
+  const pg = getKysely()
+  const r = await getRethink()
+
+  // Fetch the request that is not expired
+  const request = await pg
+    .selectFrom('DomainJoinRequest')
+    .selectAll()
+    .where('id', '=', requestId)
+    .where('expiresAt', '>', now)
+    .executeTakeFirst()
+
+  if (!request) {
+    return standardError(new Error('Request expired'))
+  }
+
+  const {createdBy, domain} = request
+
+  // Viewer must be a lead of the provided teamIds
+  const validTeamMembers = await r
+    .table('TeamMember')
+    .getAll(r.args(teamIds), {index: 'teamId'})
+    .filter({isLead: true, userId: viewerId})
+    .run()
+
+  if (!validTeamMembers.length) {
+    return standardError(new Error('Not a team leader'))
+  }
+
+  // Provided request domain should match team's organizations activeDomain
+  const leadTeams = await getTeamsByIds(validTeamMembers.map((teamMember) => teamMember.teamId))
+  const validOrgIds = await r
+    .table('Organization')
+    .getAll(r.args(leadTeams.map((team) => team.orgId)))
+    .filter({activeDomain: domain})('id')
+    .run()
+
+  if (!validOrgIds.length) {
+    return standardError(new Error('Invalid organizations'))
+  }
+
+  const validTeams = leadTeams.filter((team) => validOrgIds.includes(team.orgId))
+
+  if (!validTeams.length) {
+    return standardError(new Error('Invalid teams'))
+  }
+
+  // Avoid race conditions: we make changes based on the data fetched first, see insertNewTeamMember
```

Comment:
+1 I think we would be better served in making `insertNewTeamMember` safe by refactoring the rethink query to read and write in one go using `branch`, maybe sth. like this
```ts
r.table('TeamMember').get(teamMemberId).branch(
  r.table('TeamMember').get(teamMemberId).update({isNotRemoved: true, updatedAt: new Date()}),
  r.table('TeamMember').insert({
    ...teamMember,
    isLead: r.table('TeamMember').getAll(teamId, {index: 'teamId'}).filter({isNotRemoved: true}).count().eq(0)
  })
).run()
```

Reply:
@Dschoordsch I think this is not the only issue, then we call `adjustUserCount` and  `setUserTierForUserIds` which both also make actions based on fetched data
I feel too much refactoring would be here if it is possible